### ISSUE TITLE
Re-indent all our code

### ISSUE THREAD

**Issue by PatrickvL** [Cxbx-Reloaded/Cxbx-Reloaded#1136]
Instead of everyone applying their own indenting style to Cxbx-Reloaded code via relatively small pull requests, update our entire code base to one style, in one go, without any other change, and if possible make it so that future code automatically adheres to this style

**Comment by x1nixmzeng** [comment]
We could hook up https://clang.llvm.org/docs/ClangFormat.html - but it'd be worth defining some style guidelines first.

The downside is any third-party libraries we've imported into the repro (rather than moduled).

**Comment by PatrickvL** [comment]
True. Those would have to be excluded from automatic reformatting (except OpenXDK perhaps, since that's been modified so much, it's hardly the same thing anymore, plus, the project itself is in hibernation)

**Comment by RadWolfie** [comment]
Currently, git doesn't work with intending our code before it is pushed into repo. It seems to require some sort of extension in order to have filter option working. Only new lines are working.

I would say Clang format extension is a requirement plus AppVeyor has support for this? Or would it be TravisCI?

**Comment by x1nixmzeng** [comment]
You can setup commit hooks, which run local scripts before committing.

Personally, I wouldn't bother enforcing standards on a build machine - having an automated build fail because of style is a waste of time in my opinion :stuck_out_tongue:

**Comment by PatrickvL** [comment]
Keeping the style consistent first and foremost requires a choice for an actual style.
Automation is of later concern.

Can we list and discuss a few smart styles here?

**Comment by RadWolfie** [comment]
@x1nixmzeng Not true, it will tells us the contributor isn't following our standard which need to be fix before it can go into master.

Currently, both source code and github repo are showing both spaces and tabs making it look out of order which is not a good design. Using Clang format enforcement will help smooth the workflow better into main branch. I want to mention this, @LukeUsher did state Clang format extension is cross-platform supportive.

I think we need to change the topic to "format standard" with listing in op post.

**Comment by PatrickvL** [comment]
Let's keep it simple. First define a style. Then apply it to all our code. Then document it. And finally evaluate cost versus benefit of automation.

**Comment by RadWolfie** [comment]
We can create `.editorconfig` file with the following code below to force fix github's viewing issue between tab and spaces mixture problem.
```
[*]
indent_size = 4
```
I do not understand why github is using 8 spaces for a single tab.

`.editorconfig` file extension is supported without plugin requirement by the way.

Then we can slap in Clang format extension file to meet our standard format since there's no documentation state it is support without plugin requirement for github. We can leave `.editorconfig` file in github permanently.

**Comment by slx7R4GDZM** [comment]
>I do not understand why github is using 8 spaces for a single tab.

The standard for tabs was originally 8 spaces until recently. I think it was this way on vi and even typewriters.

**Comment by RadWolfie** [comment]
Thanks for the explanation @slx7R4GDZM! 😃

**Comment by RadWolfie** [comment]
Ident style we should be using is [OTBS](https://en.wikipedia.org/wiki/Indentation_style#Variant:_1TBS_(OTBS)) variant under K&R.

Under ideas for .editorconfig, `indent_brace_style` does not seem to apply at all with Visual Studio 2017.

Also, I notice Clang format is integrated with Visual Studio 2017. See below for example.
![image](https://user-images.githubusercontent.com/3938312/40129079-f860c852-58f8-11e8-85ed-94d6430e661d.png)

**Comment by PatrickvL** [comment]
From that same Wikipedia page, the Stroustrup style most closely describes what we're using right now (if you'd ignore tabs and indent size)

**Comment by RadWolfie** [comment]
Here's a mockup of clang format I had made. I can't find @LukeUsher's clang format file from pm months ago to compare against my code. The mockup will help enforce things better into standard coding with some flexibility of course.

```
---
AlignAfterOpenBracket: Align
AlignConsecutiveAssignments: 'true'
AlignConsecutiveDeclarations: 'true'
AlignEscapedNewlinesLeft: 'true'
AlignOperands: 'true'
AlignTrailingComments: 'true'
AllowShortCaseLabelsOnASingleLine: 'false'
AllowShortFunctionsOnASingleLine: None
AllowShortIfStatementsOnASingleLine: 'false'
AllowShortLoopsOnASingleLine: 'false'
AlwaysBreakAfterReturnType: None
AlwaysBreakBeforeMultilineStrings: 'false'
AlwaysBreakTemplateDeclarations: 'true'
BinPackParameters: 'true'
BreakBeforeBraces: Linux
IndentCaseLabels: 'true'
Language: Cpp
MaxEmptyLinesToKeep: '2'
NamespaceIndentation: All
SpaceBeforeAssignmentOperators: 'true'
SpaceBeforeParens: Never
SpaceInEmptyParentheses: 'false'
SpacesInAngles: 'false'
SpacesInCStyleCastParentheses: 'false'
SpacesInContainerLiterals: 'false'
SpacesInParentheses: 'false'
SpacesInSquareBrackets: 'false'
...
```

The reason to use the first code below is to show the chain in human readability. Only time to a have new line after if/else is another if statement. Which automatic recognize as "will check statement again", for some people like me to read and understand quickly.
```c
// Human readability
if (test == 0) {
  test = 1;
} else if (test == 4) {
  test = 2;
} else {
  test = 0;
}
if (test== 0) {
  test = 10;
}
```
vs
```c
// Not really human readability
if (test == 0) {
  test = 1;
}
else if (test == 4) {
  test = 2;
}
else {
  test = 0;
}
if (test== 0) {
  test = 10;
}
```
vs
```c
// Not very human readability plus waste of screen spaces.
if (test == 0)
{
  test = 1;
}
else if (test == 4)
{
  test = 2;
}
else
{
  test = 0;
}
if (test== 0)
{
  test = 10;
}
```

**Comment by ergo720** [comment]
Human readability is a subjective issue IMO. For example, the second case is much more readable to me than the first one.

**Comment by PatrickvL** [comment]
I've found it helpful, pretty and descriptive to put empty lines between compound statement blocks.

**Comment by RadWolfie** [comment]
Okay so only this line will need to change: `BreakBeforeBraces: Linux` to `BreakBeforeBraces: Stroustrup`. And I don't think there is a format to make if statement as new line. We'll have to monitor each other for that. Unless I have overlooked in the documentation.

**Comment by Fisherman166** [comment]
@RadWolfie What happened to this? I'd love to see this get into production as the codebase is all over the place right now.

**Comment by RadWolfie** [comment]
Each of us has various of major re-work in forked and locally, except most of them doesn't work while still has some improvement. Much as we want to put this into production, it will affect a lot of source codes while trying to make sense of what's been changed in each pull request.

This is part of a reason it's been delayed for long time and still has been.

**Comment by Margen67** [comment]
Are spaces or tabs wanted? It should be easy to re-indent everything..

**Comment by Fisherman166** [comment]
I vote 4 spaces.

Also, we should just use clang-format or a similar tool to format on pull request or commit so people don't have to think about code format.
